samcli/lib/intrinsic_resolver/intrinsic_property_resolver.py

```diff
@@ -0,0 +1,811 @@
+        Return
+        ---------
+        The simplified version of the intrinsic function. This could be a list,str,dict depending on the format required
+        """
+        if intrinsic is None:
+            raise InvalidIntrinsicException("Missing Intrinsic property in {}".format(parent_function))
+
+        if any(isinstance(intrinsic, object_type) for object_type in
+               [string_types, list, bool, int]) or intrinsic == {}:
+            return intrinsic
+
+        keys = list(intrinsic.keys())
+        key = keys[0]
+
+        if key in self.intrinsic_key_function_map:
+            intrinsic_value = intrinsic.get(key)
+            return self.intrinsic_key_function_map.get(key)(intrinsic_value)
+        elif key in self.conditional_key_function_map:
+            intrinsic_value = intrinsic.get(key)
+            return self.conditional_key_function_map.get(key)(intrinsic_value)
+
+        # In this case, it is a dictionary that doesn't directly contain an intrinsic resolver, we must recursively
+        # resolve each of it's sub properties.
+        sanitized_dict = {}
+        for key, val in intrinsic.items():
+            sanitized_key = self.intrinsic_property_resolver(key, parent_function=parent_function)
+            sanitized_val = self.intrinsic_property_resolver(val, parent_function=parent_function)
+            verify_intrinsic_type_str(sanitized_key,
+                                      message="The keys of the dictionary {} in {} must all resolve to a string"
+                                      .format(sanitized_key, parent_function))
+            sanitized_dict[sanitized_key] = sanitized_val
+        return sanitized_dict
+
+    def resolve_template(self, template=None, symbol_resolver=None, ignore_errors=False):
+        """
+        This will parse through every entry in a CloudFormation template and resolve them based on the symbol_resolver.
+        Customers can optionally ignore resource errors and default to whatever the resource provides.
+
+        Parameters
+        -----------
+        template: dict
+            A dictionary containing all the resources and the attributes of the CloudFormation template. The class does
+            not mutate the template.
+        ignore_errors: bool
+            An option to ignore
+        symbol_resolver: IntrinsicSymbolTable
+            An Instance of the IntrinsicSymbolTable object contaning resolution for attributes like Ref, Fn::GetAtt
+        Return
+        -------
+        A resolved template with all references possible simplified
+        """
+        if template:
+            self.template = template
+        if symbol_resolver:
+            self._symbol_resolver = symbol_resolver
+
+        processed_template = {}
+        for key, val in self._resources.items():
+            processed_key = self._symbol_resolver.get_translation(key) or key
```

Comment:
Should `self._symbol_resolver.get(key)` just default to returning the key back if one can't be found?

Reply:
The reason I think we should continue to return None for self._symbol_resolver.get(key) is because we may want to differentiate between keys that don't exist and symbols that references to a logical_id in the future. This also makes it feel like dict.get(), which returns a default of None.